src/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ServiceNameCollection.cs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Diagnostics;
+using System.Globalization;
+using System.Net;
+
+namespace System.Security.Authentication.ExtendedProtection
+{
+    public class ServiceNameCollection : ReadOnlyCollectionBase
+    {
+        public ServiceNameCollection(ICollection items)
+        {
+            if (items == null)
+            {
+                throw new ArgumentNullException("items");
+            }
+
+            // Normalize and filter for duplicates.
+            foreach (string serviceName in items)
+            {
+                AddIfNew(InnerList, serviceName);
+            }
+        }
+
+        public ServiceNameCollection Merge(string serviceName)
+        {
+            ArrayList newServiceNames = new ArrayList();
```

Comment:
Can these methods be upgraded to something newer than ArrayList?

In particular, the `Merge(IEnumerable)` seems like it could benefit from a HashSet, unless the order is guaranteed (even then, a HashSet for duplicate detection and a List for ordering seems +EV)